### ISSUE TITLE
Does not work anymore with latest routeros release

### ISSUE THREAD

**Issue by sky-home** [instalator/ioBroker.mikrotik#23]
Since upgrade of routeros release >= 6.45 the plugin does not conenct to the routerboard anymore:
Instance "led" indocator is yellow and mouseover says "connected to Mikrotik: False"

When trying to change an interface error os reported:

- host.raspi-iobroker | 2019-07-22 10:23:20.930 | error | instance system.adapter.mikrotik.0 terminated with code 6 (uncaught exception)
- -- | -- | -- | --
- Caught | 2019-07-22 10:23:20.930 | error | by controller[0]:     at Decoder.Emitter.emit  (/opt/iobroker/node_modules/socket.io-parser/node_modules/component-emitter/index.js:134:20)
- Caught | 2019-07-22 10:23:20.929 | error | by controller[0]:     at Decoder.<anonymous> (/opt/iobroker/node_modules/component-bind/index.js:21:15)
- Caught | 2019-07-22 10:23:20.929 | error | by controller[0]:     at Manager.ondecoded (/opt/iobroker/node_modules/socket.io-client/lib/manager.js:332:8)
- Caught | 2019-07-22 10:23:20.929 | error | by controller[0]:     at Manager.Emitter.emit (/opt/iobroker/node_modules/component-emitter/index.js:133:20)
- Caught | 2019-07-22 10:23:20.929 | error | by controller[0]:     at Manager.<anonymous> (/opt/iobroker/node_modules/component-bind/index.js:21:15)
- Caught | 2019-07-22 10:23:20.928 | error | by controller[0]:     at Socket.onpacket (/opt/iobroker/node_modules/socket.io-client/lib/socket.js:236:12)
- Caught | 2019-07-22 10:23:20.928 | error | by controller[0]:     at Socket.onack (/opt/iobroker/node_modules/socket.io-client/lib/socket.js:312:9)
- Caught | 2019-07-22 10:23:20.928 | error | by controller[0]:     at Socket.<anonymous> (/opt/iobroker/node_modules/iobroker.mikrotik/main.js:93:21)
- Caught | 2019-07-22 10:23:20.927 | error | by controller[0]:     at GetCmd (/opt/iobroker/node_modules/iobroker.mikrotik/main.js:120:5)
- Caught | 2019-07-22 10:23:20.927 | error | by controller[0]:     at SetCommand (/opt/iobroker/node_modules/iobroker.mikrotik/main.js:125:10)
- Caught | 2019-07-22 10:23:20.926 | error | by controller[0]: TypeError: Cannot read property 'write' of undefined
- mikrotik.0 | 2019-07-22 10:23:20.890 | info | cleaned everything up...

I assume this error is releated to the change in API authentification - release notes from Mikrotik:

What's new in 6.45.1 (2019-Jun-27 10:23):

Important note!!!
Due to removal of compatibility with old version passwords in this version, downgrading to any version prior to v6.43 (v6.42.12 and older) will clear all user passwords and allow password-less authentication. Please secure your router after downgrading.
**Old API authentication method will also no longer work, see documentation for new login procedure:
https://wiki.mikrotik.com/wiki/Manual:API#Initial_login**

**Comment by instalator** [comment]
https://github.com/f5eng/mikronode-ng/issues/22

**Comment by sky-home** [comment]
**@instalator: Thank you!**

How to fix: Change _/opt/iobroker/node_modules/mikronode-ng/lib/connections.js_ like this:

   **...                
   // lc.write('/login'); - OLD Mikrotik firmware
   // _(this).status = 'Waiting for challenge'; - OLD Mikrotik firmware
   lc.write('/login', { "=name" : _(this).user, "=password" : _(this).password });
   _(this).status = 'Waiting for Response';
   ...**

After _restart of Mikrotik iobroker instance_ the connection was established successfully.

I would prefer to have mikronode-ng (+mikrotik iobroker instance) fixed without the need to edit connection.js.

**Comment by instalator** [comment]
If the author does not update the module, then I will fork.

**Comment by instalator** [comment]
fixed 1.0.5

**Comment by RodrigoRodriguesX10** [comment]
I had the same issue, but I don't want to write into connections.js file, it's not recommended for development. 
So I just change the reference of the function "localStateMachine", like this:

```js
var mikrotik = require("mikronode-ng");
var _ = require('private-parts').createKey();
module.exports = {
    loginPost645: function (config, callback) {
        var connection = mikrotik.getConnection(config.address, config.user, config.password, { port: config.port || 8728 });
        connection.loginStateMachine = function loginStateMachine() {
            _(connection).closeOnDone = false;
            var lc = this.openChannel('login');
            lc.closeOnDone = false;
            _(connection).status = 'Sending Login';
            lc.write('/login');
            _(connection).status = 'Waiting for challenge';
            var _this = this;
            lc.on('trap', function loginTrap(data) {
                _(_this).status = 'Received Trap';
                _this.emit('trap', data);
                this.close(true);
                _(_this).closeOnDone = true;
                return callback(data);
            });

            lc.on('done', function loginDone(data) {
                switch (_(_this).status) {
                    case 'Waiting for challenge':
                        lc.write('/login', { "=name": _(connection).user, "=password": _(connection).password });
                        _(_this).status = 'Waiting for Response';
                        return;
                    case 'Waiting for Response':
                        _(_this).status = 'Received Response';
                        this.close(true);
                        _(connection).status = 'Connected';
                        _(connection).connected = true;
                        connection.loginStateMachine = old;
                        return cb(null, connection);
                }
            });
        };
        connection.on("error", function (data) {
            console.log("Error during connection:: ", data);
            return cb(data);
        });
        return connection.connect();
    }
}
```

I know it is rewriting the code, but I keep changes in my project files without update module's file. Also I can call loginPost645 only when the mikrotik is >= 6.45 version, for anyone else I keep using the module's implementation.
